File: c/hamming/hamming.c
#include "hamming.h"

int compute(const char *lhs, const char *rhs) {
    int distance = 0;
    int offset = 0;

    while (*(lhs + offset) != '\0' || *(rhs + offset) != '\0') {
        char left_char = *(lhs + offset);
        char right_char = *(rhs + offset);

        if (left_char == '\0' || right_char == '\0') {
            return -1;
        }

        if (left_char != right_char) {
            distance++;
        }

        offset += sizeof(char);
    }

    return distance;
}

File: lua/hamming/hamming.lua
local Hamming = {}

function Hamming.compute(a, b)
    if #a ~= #b then
        error("strands must be of equal length")
    end

    local distance = 0

    for i = 1, #a do
        local c1 = string.sub(a, i, i)
        local c2 = string.sub(b, i, i)

        if c1 ~= c2 then
            distance = distance + 1
        end
    end

    return distance
end

return Hamming
